<?xml version="1.0"?>
<robot name="arduino_mecanum" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- 你可以沒有幾何，只要 ros2_control 區塊就能跑 -->
  <ros2_control name="ArduinoSystem" type="system">
    <hardware>
      <plugin>bennis_hardware_interface/BennisHardware</plugin>

      <!-- 這些參數要對應你在 configure() 讀的 hardware_parameters -->
      <param name="device">/dev/ttyACM0</param>
      <param name="baud_rate">115200</param>
      <param name="timeout">50</param>
      <param name="loop_rate">50</param>
      <param name="enc_counts_per_rev">2048</param>
      <param name="left_wheel_name">left_wheel_joint</param>
      <param name="right_wheel_name">right_wheel_joint</param>
      <param name="left_back_wheel_name">left_back_wheel_joint</param>
      <param name="right_back_wheel_name">right_back_wheel_joint</param>
    </hardware>

    <joint name="left_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface   name="velocity"/>
      <state_interface   name="position"/>
    </joint>

    <joint name="right_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface   name="velocity"/>
      <state_interface   name="position"/>
    </joint>

    <joint name="left_back_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface   name="velocity"/>
      <state_interface   name="position"/>
    </joint>
    <joint name="right_back_wheel_joint">
      <command_interface name="velocity"/>
      <state_interface   name="velocity"/>
      <state_interface   name="position"/>
    </joint>
  </ros2_control>
</robot>
